zypp/KeyManager.cc

```diff
@@ -216,114 +219,153 @@ std::list<std::string> KeyManagerCtx::Impl::readSignaturesFprsOptVerify( const P
+      ZYPP_THROW( GpgmeException( "gpgme_ctx_set_engine_info", err ) );
   }
 
-  shared_ptr<KeyManagerCtx> me( new KeyManagerCtx());
-  me->_pimpl->_ctx = ctx;
-  return me;
+  return ret;
 }
 
-bool KeyManagerCtx::setHomedir(const Pathname &keyring_r)
+Pathname KeyManagerCtx::homedir() const
 {
+  Pathname ret;
+  if ( gpgme_engine_info_t enginfo = gpgme_ctx_get_engine_info( _pimpl->_ctx ) )
+    ret = enginfo->home_dir;
+  return ret;
+}
 
-  /* get engine information to read current state*/
-  gpgme_engine_info_t enginfo = gpgme_ctx_get_engine_info(_pimpl->_ctx);
-  if (!enginfo)
-    return false;
+std::list<PublicKeyData> KeyManagerCtx::listKeys()
+{
+  std::list<PublicKeyData> ret;
+  GpgmeErr err = GPG_ERR_NO_ERROR;
 
-  GpgmeErr err = gpgme_ctx_set_engine_info(
-        _pimpl->_ctx,
-        GPGME_PROTOCOL_OpenPGP,
-        enginfo->file_name,
-        keyring_r.c_str());
+  // Reset gpgme_keylist_mode on return!
+  AutoDispose<gpgme_keylist_mode_t> guard { gpgme_get_keylist_mode( _pimpl->_ctx ), bind( &gpgme_set_keylist_mode, _pimpl->_ctx, _1 ) };
+  // Let listed keys include signatures (required if PublicKeyData are created from the key)
+  if ( (err = gpgme_set_keylist_mode( _pimpl->_ctx, GPGME_KEYLIST_MODE_LOCAL | GPGME_KEYLIST_MODE_SIGS )) != GPG_ERR_NO_ERROR ) {
+    ERR << "gpgme_set_keylist_mode: " << err << endl;
+    return ret;
+  }
 
-  if (err != GPG_ERR_NO_ERROR) {
-    ERR << "Unable to set homedir " << err << endl;
-    return false;
+  if ( (err = gpgme_op_keylist_start( _pimpl->_ctx, NULL, 0 )) != GPG_ERR_NO_ERROR ) {
+    ERR << "gpgme_op_keylist_start: " << err << endl;
+    return ret;
+  }
+  // Close list operation on return!
+  AutoDispose<gpgme_ctx_t> guard2 { _pimpl->_ctx, &gpgme_op_keylist_end };
+
+  AutoDispose<gpgme_key_t> key { nullptr, &gpgme_key_release };
+  for ( ; gpgme_op_keylist_next( _pimpl->_ctx, &(*key) ) == GPG_ERR_NO_ERROR; key.getDispose()( key ) ) {
+    PublicKeyData data { PublicKeyData::fromGpgmeKey( key ) };
+    if ( data )
+      ret.push_back( data );
   }
 
-  return true;
+  return ret;
 }
 
-Pathname KeyManagerCtx::homedir() const
+#if 1
+std::list<PublicKeyData> KeyManagerCtx::readKeyFromFile( const Pathname & keyfile_r )
 {
-  gpgme_engine_info_t enginfo = gpgme_ctx_get_engine_info(_pimpl->_ctx);
-  if (!enginfo)
-    return Pathname();
+  // bsc#1140670: GPGME does not support reading keys from a keyfile using
+  // gpgme_data_t and gpgme_op_keylist_from_data_start. Despite GPGME_KEYLIST_MODE_SIGS
+  // the signatures are missing, but we need them to create proper PublicKeyData objects.
+  // While this is not resolved, we read into a temp. keyring. Impl::_volatile helps
+  // to detect whether we can clear and import into the current context or need to
+  // create a temp. one.
+  std::list<PublicKeyData> ret;
+
+  if ( _pimpl->_volatile ) {
+    // in a volatile context we can simple clear the keyring...
+    filesystem::clean_dir( homedir() );
```

Comment:
I wonder if its a good idea to do this. It kind of is abusing undefined behaviour of what gpgme does when the homedir() is cleared. Upgrading gpgme could change the behaviour and break that part. 
Did you try to use a statically created KeyManagerCtx for all the volatile cases and purge it instead after the op is done?

Reply:
> All cryptographic operations in GPGME are performed within a context, which contains the internal state of the operation as well as configuration parameters. By using several contexts you can run several cryptographic operations in parallel, with different configuration. 

GPGME is and must be prepared for concurrent access to the homedir from the same process as well as from others. Also any GPGME operation creates a missing keyring on demand in case it is needed (not when a context is created). The behavior is IMO not that undefined (and not much different from using the keyring dir the 1st time).

Deleting all the imported keys again is also possible, but far slower.
(and as soon as bsc#1140670 is fixed, we don't need this hack at all)